pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -409,11 +424,133 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+        key_file = os.environ["SIGNING_KEY"]
+        uri = constants.UMB_BROKERS[self.signing_env]
+        sig_keyname = "redhatrelease2" if client_type == 'ocp' else "beta2"
+        self._logger.info("About to sign artifacts with key %s", sig_keyname)
+        json_digest_sig_dir = self._working_dir / "json_digests"
+        message_digest_sig_dir = self._working_dir / "message_digests"
+        base_to_mirror_dir = self._working_dir / "to_mirror/openshift-v4"
+
+        lock_name = f'signing-lock'
+        lock_policy = locks.LOCK_POLICY['compose']
+        lock_manager = locks.new_lock_manager(
+            internal_lock_timeout=lock_policy['lock_timeout'],
+            retry_count=lock_policy['retry_count'],
+            retry_delay_min=lock_policy['retry_delay_min']
+        )
+        async with await lock_manager.lock(lock_name):
+            async with AsyncSignatory(uri, cert_file, key_file, sig_keyname=sig_keyname) as signatory:
+                tasks = []
+                for _, release_info in release_infos.items():
+                    pullspec = release_info["image"]
+                    digest = release_info["digest"]
+                    sig_file = json_digest_sig_dir / f"{digest.replace(':', '=')}" / "signature-1"
+                    tasks.append(self._sign_json_digest(signatory, release_info["metadata"]["version"], pullspec, digest, sig_file))
+                for message_digest in message_digests:
+                    input_path = base_to_mirror_dir / message_digest
+                    if not input_path.is_file():
+                        raise IOError(f"Message digest file {input_path} doesn't exist or is not a regular file")
+                    sig_file = message_digest_sig_dir / f"{message_digest}.gpg"
+                    tasks.append(self._sign_message_digest(signatory, release_name, input_path, sig_file))
+                await asyncio.gather(*tasks)
+
+        self._logger.info("All artifacts have been successfully signed.")
+        self._logger.info("Publishing signatures...")
+        tasks = [
+            self._publish_json_digest_signatures(json_digest_sig_dir),
+            self._publish_message_digest_signatures(message_digest_sig_dir),
+        ]
+        await asyncio.gather(*tasks)
+        self._logger.info("All signatures have been published.")
+
+    async def _sign_json_digest(self, signatory: AsyncSignatory, release_name: str, pullspec: str, digest: str, sig_path: Path):
+        """ Sign a JSON digest claim
+        :param signatory: Signatory
+        :param pullspec: Pullspec of the payload
+        :param digest: SHA256 digest of the payload
+        :param sig_path: Where to save the signature file
+        """
+        self._logger.info("Signing json digest for payload %s with digest %s...", pullspec, digest)
+        sig_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(sig_path, "wb") as sig_file:
+            if self.runtime.dry_run:
```

Comment:
I would move this check on the very beginning of the function. We probably don't need to `mkdir`, and even if the context manager will probably close the `sig_file` on a return statement, we can probably avoid creating it in the first place

Reply:
It was designed like this only because it was easier to test and a lot of functions in standard libraries accepting stream-like objects.

Nowadays forgetting to close a file is no longer an issue since Python uses reference counting and it will close the file when the reference count goes to `0`.